### ISSUE TITLE
ads-github-fetch-all-upstreams: default to triangular workflow, but allow override

### ISSUE THREAD

**Issue by salewski** [salewski/ads-github-tools#39]
This is related to:

   * issue #38 ("ads-github-fetch-all-upstreams: chokes attempting to clone non-fork repos")
   * issue #46 ("git config: honor 'hub.upstreamremote' and 'hub.forkremote', if set")
----

The underlying `git hub` tool (see [sociomantic-tsunami/git-hub](https://github.com/sociomantic-tsunami/git-hub)) that is used when cloning repositories changed behavior in that project's [version 1.0.0](https://github.com/sociomantic-tsunami/git-hub/releases/tag/v1.0.0) to make the triangular workflow the default, and also to name the git remote associated with your GitHub fork `fork` rather than `origin`. Because our `ads-github-fetch-all-upstreams` tool is not explicitly requesting the non-triangular setup (it used to be the default), repos cloned recently are setup with the triangular workflow.

~~The `ads-github-fetch-all-upstreams` tool should be updated to require `git-hub` 1.0.0 or newer, and should explicitly control it's behavior to _not_ use a triangular setup (the author's preference). A new command line option should be required to allow the user to request that a triangular setup be used.~~

~~See also issue #46 on the higher-level rationale for prefering a non-triangular workflow as the default. In short, a user cloning a GitHub fork directly on the command line using `git(1)` would, by default, have a local repo configured to pull and push to the single "origin" remote. We want to keep the default behavior of the `ads-github-tools` consistent with that default behavior (but allow the user to override the default behavior, as outlined above).~~

**UPDATE (2020-10-12):** Since this issue was originally written, the author has changed his opinion on whether the triangular workflow should be the default. The `ads-github-fetch-all-upstreams` program should continue to use a triangular workflow by default, but should provide a `--no-triangular` option to allow the user to choose the old behavior. For symmetry, a `--triangular` option can be provided to make clear the desired behavior, even though that will be the default.

So why the change? Two reasons:
   1. In practice, when using a repo cloned for a non-triangular workflow, the author always changes the default branch to track the branch of the same name from the "upstream" repo rather than the "origin" repo. Specifying that the newly clone repo should be configured to support a triangular workflow would make this step unnecessary; it would already be configured that way.
   2. The author does not rely very much on the default pull and push targets; always specifying the remotes explicitly (`git remote update foo` and `git push foo whatever`) make the defaults that would otherwise be used irrelevant, practically speaking.

Note that (1) negates the rationale for sticking more closely to the default configuration obtained by doing a direct `git clone` of a repo. Such a config has only one remote ("origin"). Once an "upstream" remote is added, it is generally more useful to ask "how far away is my default branch than that of upstream?" than asking the same relative to the origin remote.

**Comment by salewski** [comment]
This issue as been fixed for 'ads-github-tools-0.3.1'

[The blurb about it](../blob/master/NEWS#issue-39-ads-github-fetch-all-upstreams-default-to-triangular-workflow-but-allow-override) from our ['NEWS'](../blob/master/NEWS) file:

```
*** issue 39: 'ads-github-fetch-all-upstreams: default to triangular workflow, but allow override'

    https://github.com/salewski/ads-github-tools/issues/39

    Consistent with the 'ads-github-tools-0.3.0' release, when initially
    cloning a repo the 'ads-github-fetch-all-upstreams' program continues to
    use a "triangular" workflow by default.

    The manpage entry explains:
    #+BEGIN_QUOTE
       --triangular
       --no-triangular
           In newly cloned repos, use (or avoid using) a Git remote
           configuration suitable for supporting a triangular workflow.

           Among other things, a triangular config will cause the default
           branch in the local working directory to be configured to track the
           branch of the same name from the upstream remote.

           For a non-triangular configuration, the default branch in the local
           working directory will be configured to track the branch of the
           same name from the origin remote.

           See git-hub(1) for details.

           If neither "--triangular" nor "--no-triangular" is specified, then
           the "hub.triangular" setting from git-config(1) (scopes "global"
           and "system" only) is consulted, and honored if present. Finally,
           if neither command line options nor configuration specifies a
           preference, then a triangular configuration is the default.

           Note: Whether a triangular or non-triangular configuration is used
           is purely a matter of preference for the user (if any) that will be
           working with the newly cloned repo; it has no bearing on the
           functionality of the "ads-github-fetch-all-upstreams" program.
    #+END_QUOTE
```

**Comment by salewski** [comment]
Reopening issue.

The `--triangular` and `--no-triangular` options work as advertized, but the default behavior is not what is documented above -- the `fetch-all-upstreams` program is defaulting to using a non-triangular workflow. It should be defaulting to using a triangular workflow.

**Comment by salewski** [comment]
The updated fix for this issue will be part of the upcoming 'ads-github-tools-0.3.3' release.

[The blurb about it](../blob/master/NEWS#issue-39-ads-github-fetch-all-upstreams-default-to-triangular-workflow-but-allow-override-reopened) from our ['NEWS'](../blob/master/NEWS) file:

```
*** issue 39: "ads-github-fetch-all-upstreams: default to triangular workflow, but allow override" (reopened)

    https://github.com/salewski/ads-github-tools/issues/39

    This issue was originally noted as fixed in ads-github-tools 0.3.1, but a
    bug was found with the implementation. The '--triangular' and
    '--no-triangular' options worked as advertized, but the default behavior
    of the app was inconsistent with what was documented. The documentation
    indicated that the default behavior was that the tool would use a
    triangular configuration when cloning a repo if the user did not otherwise
    specify a preference, and that was indeed the intention.

    The program's behavior has been corrected to be consistent with the
    documentation in this regard; by default a triangular configuration will
    be used.
```
